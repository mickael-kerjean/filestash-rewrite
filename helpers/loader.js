import { get as getRelease } from "../pages/adminpage/model_release.js";

let version = null;

export async function loadScript(url) {
    const $script = document.createElement("script");
    $script.setAttribute("src", url);
    document.head.appendChild($script);
    return new Promise((done) => {
        $script.onload = done;
        $script.onerror = done;
    });
}

export async function CSS(baseURL, ...arrayOfFilenames) {
    const sheets = await Promise.all(arrayOfFilenames.map((filename) => loadSingleCSS(baseURL, filename)));
    return sheets.join("\n\n");
}

async function loadSingleCSS(baseURL, filename) {
    const res = await fetch(baseURL.replace(/(.*)\/[^\/]+$/, "$1/") + `${filename}?version=${version}`, {
        cache: "force-cache",
    });
    if (res.status !== 200) return `/* ERROR: ${res.status} */`;
    else if (!res.headers.get("Content-Type")?.startsWith("text/css")) return `/* ERROR: wrong type, got "${res.headers?.get("Content-Type")}"*/`;
    return await res.text();
}

export async function init() {
    const info = await getRelease().toPromise();
    version = info.version;
}
