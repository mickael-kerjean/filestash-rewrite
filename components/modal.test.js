import { CSS } from "../helpers/loader.js";
jest.mock("../helpers/loader.js");

describe("component::modal", () => {
    beforeEach(() => {
        global.customElements = window.customElements;
        CSS.mockImplementation(() => Promise.resolve(""));
        window.setTimeout = (cb) => cb()
    });
    afterEach(() => {
        document.body.innerHTML = "";
        global.customElements = { define: jest.fn() };
    });
    it("external API", () => {
        const modal = require("./modal.js").default;
        expect(typeof modal.alert).toBe("function");
    });

    it("is an web component", async () => {
        const modal = require("./modal.js").default;
        document.body.innerHTML = `
            <div>
                <component-modal></component-modal>
            </div>
        `;
        await modal.alert(document.createElement("div"), {});
        await nextTick()
        expect(document.body).toMatchSnapshot();
    });

    it("throws when components does not exists", () => {
        const modal = require("./modal.js").default;
        document.body.innerHTML = `<div></div>`;
        expect(() => {
            modal.alert(document.createElement("div"), {});
        }).toThrow();
    });
});
