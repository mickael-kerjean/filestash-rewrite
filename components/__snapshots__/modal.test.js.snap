// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component::modal is an web component 1`] = `
<body>
  
            
  <div>
    
                
    <component-modal>
      <div
        class="component_modal"
        id="modal-box"
      >
        
        
        <style />
        
        
        <div
          style="opacity: 0; margin: 0px auto 0px auto;"
        >
          
            
          <div
            class="component_popup"
          >
            
                
            <div
              class="popup--content"
            >
              
                    
              <div
                class="modal-message"
                data-bind="body"
              >
                <div />
              </div>
              
                
            </div>
            
                
            <div
              class="buttons"
            >
              
                    
              <button
                class="emphasis"
                type="submit"
              >
                OK
              </button>
              
                
            </div>
            
            
          </div>
          
        
        </div>
        
    
      </div>
    </component-modal>
    
            
  </div>
  
        
</body>
`;
